admin/server/device_code.go

```diff
@@ -0,0 +1,315 @@
+		return
+	}
+	scopes := strings.Split(values.Get("scope"), " ")
+	if len(scopes) == 0 {
+		http.Error(w, "scope is required", http.StatusBadRequest)
+		return
+	}
+	if len(scopes) > 1 || scopes[0] != "full_account" {
+		http.Error(w, "invalid scope", http.StatusBadRequest)
+		return
+	}
+	authCode, err := generateDeviceAndUserCode()
+	if err != nil {
+		s.logger.Error("failed to generate device and user code", zap.Error(err))
+		internalServerError(w, err)
+		return
+	}
+	authCode.ClientID = clientID
+	err = s.admin.DB.InsertAuthCode(r.Context(), authCode)
+	if err != nil {
+		s.logger.Error("failed to insert auth code", zap.Error(err))
+		internalServerError(w, err)
+		return
+	}
+	verificationURI := s.conf.DeviceVerificationHost + "/oauth/device"
+	resp := DeviceCodeResponse{
+		DeviceCode:              authCode.DeviceCode,
+		UserCode:                authCode.UserCode,
+		VerificationURI:         verificationURI,
+		VerificationCompleteURI: verificationURI + "?user_code=" + authCode.UserCode,
+		ExpiresIn:               tokenExpirationSeconds,
+		PollingInterval:         5,
+	}
+	e := json.NewEncoder(w)
+	e.SetIndent("", "  ")
+	err = e.Encode(resp)
+	if err != nil {
+		s.logger.Error("failed to encode response", zap.Error(err))
+		internalServerError(w, err)
+	}
+}
+
+// handleUserCodeConfirmation handles the user code confirmation page. The user code is displayed
+// to the user and they are asked to confirm that they want to authorize the device. If the user
+// confirms, the device code is marked as approved in the server's device code store.
+func (s *Server) handleUserCodeConfirmation(w http.ResponseWriter, r *http.Request, pathParams map[string]string) {
+	userCode := r.URL.Query().Get("user_code")
+	if userCode == "" {
+		http.Error(w, "user_code is required", http.StatusBadRequest)
+		return
+	}
+	confirmation := r.URL.Query().Get("code_confirmed")
+	if confirmation == "" {
+		http.Error(w, "no code confirmation", http.StatusBadRequest)
+		return
+	}
+	sess, err := sessions.NewCookieStore([]byte(s.conf.SessionSecret)).Get(r, authSessionName)
+	if err != nil {
+		http.Error(w, fmt.Sprintf("failed to get session: %s", err), http.StatusInternalServerError)
+		return
+	}
+
+	var profiles map[string]interface{}
+	profile := sess.Values["profile"].([]byte)
+	err = json.Unmarshal(profile, &profiles)
+	if err != nil {
+		s.logger.Error("failed to unmarshal profile", zap.Error(err))
+		internalServerError(w, err)
+		return
+	}
+	userID := profiles["sub"].(string)
+	if userID == "" {
+		s.logger.Error("no subject in profile: " + string(profile))
+		internalServerError(w, err)
+		return
+	}
+	authCode, err := s.admin.DB.GetByUserCode(r.Context(), userCode)
+	if err != nil {
+		if errors.Is(err, sql.ErrNoRows) {
+			http.Error(w, fmt.Sprintf("no such user code: %s found", userCode), http.StatusBadRequest)
+			return
+		}
+		s.logger.Error("failed to get auth code for userCode: "+userCode, zap.Error(err))
+		internalServerError(w, err)
+		return
```

Comment:
my thinking was if user code is incorrect then we will get `sql.ErrNoRows` so we return `StatusBadRequest` however if something else goes wrong like database connectivity issues etc. then we should log it and return internal server error

Reply:
I see that, that makes sense. However, we should still not log at every internal error as described above – better to add a middleware. Will prioritize it next sprint, so better keep this clean to prevent needing to refactor all these later as well.